### ISSUE TITLE
Height of vertical scrollbar is not accurately calculated when attached to LazyColumn

### ISSUE THREAD

**Issue by igordmn** [JetBrains/compose-jb#1211]
Compose 1.0.0-alpha4-build366

Things get worse if LazyColumn has contentPadding, in this case the scrollbar can go beyond the window boundaries. I think that accurateness of the computations also depends on the number of items and/or their heights. The paddings of the scrollbar itself can add to the issue. The reproducer is attached.

```
import androidx.compose.desktop.DesktopMaterialTheme
import androidx.compose.foundation.LocalScrollbarStyle
import androidx.compose.foundation.VerticalScrollbar
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.rememberScrollbarAdapter
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Window
import androidx.compose.ui.window.application

@Composable
fun App() {
    DesktopMaterialTheme {
        Box(Modifier.fillMaxSize()) {
            val columnState = rememberLazyListState()
            LazyColumn(
                Modifier.fillMaxSize(),
                state = columnState,
                contentPadding = PaddingValues(vertical = 24.dp),
            ) {
                for (step in 1..11) {
                    item {
                        Text(step.toString(), Modifier.height((step * 9).dp))
                    }
                }
            }
            VerticalScrollbar(
                modifier = Modifier
                    .align(Alignment.TopCenter)
                    .padding(PaddingValues(vertical = 3.dp)),
                adapter = rememberScrollbarAdapter(columnState),
                style = LocalScrollbarStyle.current.copy(
                    unhoverColor = MaterialTheme.colors.onBackground.copy(alpha = 0.3f),
                )
            )
            VerticalScrollbar(
                modifier = Modifier
                    .align(Alignment.TopCenter)
                    .offset(x = 24.dp)
                    .padding(PaddingValues(vertical = 0.dp)),
                adapter = rememberScrollbarAdapter(columnState),
                style = LocalScrollbarStyle.current.copy(
                    unhoverColor = MaterialTheme.colors.onBackground.copy(alpha = 0.3f),
                )
            )
        }
    }
}

fun main() = application {
    Window(onCloseRequest = ::exitApplication) {
        App()
    }
}
```

https://user-images.githubusercontent.com/5963351/135318468-25f79d7f-2e63-4112-9d6b-b9f3cc13dc6b.mov

**Comment by m-sasha** [comment]
There may be several separate issues here, so I'd like to separate them.

One issue is that the `ScrollBarAdapter` for `LazyListState` doesn't take into account the `LazyList`s `contentPadding`. Here's a reproducer focusing on that issue specifically:

```
package androidx.compose.ui

import androidx.compose.foundation.VerticalScrollbar
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.rememberScrollbarAdapter
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Window
import androidx.compose.ui.window.application
import androidx.compose.ui.window.rememberWindowState


fun main() = application {
    Window(
        onCloseRequest = ::exitApplication,
        state = rememberWindowState(width = 300.dp, height = 400.dp)
    ) {
        LazyScrollable()
    }
}


@Composable
fun LazyScrollable() {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(color = Color.White)
    ) {

        val state = rememberLazyListState()

        LazyColumn(
            modifier = Modifier.fillMaxSize(),
            state = state,
            contentPadding = PaddingValues(vertical = 100.dp)
        ) {
            items(5) { x ->
                TextBox("Item #$x", x)
            }
        }
        VerticalScrollbar(
            modifier = Modifier.align(Alignment.CenterEnd).fillMaxHeight(),
            adapter = rememberScrollbarAdapter(
                scrollState = state,
            )
        )
    }
}


@Composable
fun TextBox(text: String = "Item", index: Int) {
    Box(
        modifier = Modifier
            .height(130.dp)
            .fillMaxWidth()
            .background(
                color = Color(
                    red = (index * 50).mod(255),
                    green = (index * 100).mod(255),
                    blue = (index * 150).mod(255),
                    alpha = 128
                )
            ),
        contentAlignment = Alignment.CenterStart
    ) {
        Text(text = text)
    }
}
```

https://user-images.githubusercontent.com/5230206/211145019-bab2f59b-ec34-4b20-803d-11ef7e7a3871.mp4

As you can see, dragging the scrollbar to the bottom doesn't cause the content to scroll to the bottom.

**Comment by m-sasha** [comment]
After fixing https://github.com/JetBrains/compose-jb/issues/2604 there is still another issue here where the scrollbar sometimes doesn't scroll all the way to the bottom (even without scrollbar padding) in the presence of non-zero content padding.

Below is a reproducer that shows the issue well, with a large gap between the bottom of the scrollbar's thumb and the bottom of the scrollbar itself. The behavior is very sensitive to the heights of the items and the height of the viewport.

```
import androidx.compose.foundation.VerticalScrollbar
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.rememberScrollbarAdapter
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Window
import androidx.compose.ui.window.application
import androidx.compose.ui.window.rememberWindowState


fun main() = application {
    Window(
        onCloseRequest = ::exitApplication,
        state = rememberWindowState(width = 300.dp, height = 380.dp)
    ) {
        LazyScrollable()
    }
}


@Composable
fun LazyScrollable() {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(color = Color.White)
    ) {
        val state = rememberLazyListState()
        val heights = listOf(103.dp, 201.dp, 75.dp)

        LazyColumn(
            modifier = Modifier.fillMaxSize(),
            state = state,
            contentPadding = PaddingValues(top = 100.dp, bottom = 200.dp),
        ) {
            items(heights.size) { x ->
                TextBox("Item #$x", x, heights[x])
            }
        }

        VerticalScrollbar(
            modifier = Modifier.align(Alignment.CenterEnd).fillMaxHeight(),
            adapter = rememberScrollbarAdapter(
                scrollState = state,
            )
        )
    }
}


@Composable
fun TextBox(text: String = "Item", index: Int, height: Dp) {
    Box(
        modifier = Modifier
            .height(height)
            .fillMaxWidth()
            .background(
                color = Color(
                    red = (index * 50).mod(255),
                    green = (index * 100).mod(255),
                    blue = (index * 150).mod(255),
                    alpha = 128
                )
            ),
        contentAlignment = Alignment.CenterStart
    ) {
        Text(text = text)
    }
}
```

https://user-images.githubusercontent.com/5230206/211155053-3fdb159a-feb9-4c26-805f-b94e595a2ff0.mp4

There is also a large jump "back" that occurs when scrolling near the bottom.

**Comment by m-sasha** [comment]
Hmm, there seems to be another issue here. If we set padding on a scrollbar, its height becomes different from the viewport height. `SliderAdapter`, however, appears to assume that they're always the same - it uses `containerSize` as if it's both.

Reproducer:

```
import androidx.compose.foundation.VerticalScrollbar
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.rememberScrollbarAdapter
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Window
import androidx.compose.ui.window.application
import androidx.compose.ui.window.rememberWindowState


fun main() = application {
    Window(
        onCloseRequest = ::exitApplication,
        state = rememberWindowState(width = 300.dp, height = 380.dp)
    ) {
        Scrollable()
    }
}


@Composable
fun Scrollable() {
    Box(
        modifier = Modifier.fillMaxSize()
    ){
        val listState = rememberLazyListState()
        LazyColumn(
            state = listState
        ) {
            items(20){
                TextBox("Item #$it")
                if (it < 20) {
                    Spacer(modifier = Modifier.height(5.dp))
                }
            }
        }
        VerticalScrollbar(
            modifier = Modifier
                .align(Alignment.CenterEnd)
                .fillMaxHeight()
                .padding(vertical = 50.dp)
                .background(Color.Blue.copy(alpha=0.5f)),
            adapter = rememberScrollbarAdapter(listState)
        )
    }
}

@Composable
fun TextBox(text: String = "Item") {
    Box(
        modifier = Modifier.height(32.dp)
            .fillMaxWidth()
            .background(color = Color(200, 0, 0, 20))
            .padding(start = 10.dp),
        contentAlignment = Alignment.CenterStart
    ) {
        Text(text = text)
    }
}
```

https://user-images.githubusercontent.com/5230206/211157560-b5cee9a4-43f0-40d5-8561-3be42b8e9eb6.mp4

**Comment by m-sasha** [comment]
I think that
- [Make ScrollbarAdapter for LazyList take contentPadding into account.](https://github.com/JetBrains/androidx/pull/365)
- [Declare and implement a new v2.ScrollbarAdapter](https://github.com/JetBrains/androidx/pull/368)
- [Compute LazyScrollbarAdapter.scrollOffset via layoutInfo.visibleItemsInfo instead of firstVisibleItemIndex](https://github.com/JetBrains/androidx/pull/372)
together fix all the issues seen in this ticket.

**Comment by igordmn** [comment]
Let's close it then. Will be fixed in Compose 1.4.0 (because It is better to not include this fix into 1.3.0: https://github.com/JetBrains/androidx/pull/368)

**Comment by okushnikov** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/CMP-5241">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
